Boost: Build fails with boost version 1.72.0

```
/usr/include/boost/asio/detail/bind_handler.hpp:170:5: error: no matching function for call to object of type 'const boost::asio::detail::write_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::mutable_buffers_1, const boost::asio::mutable_buffer *, boost::asio::detail::transfer_all_t, std::tr1::_Bind<std::tr1::_Mem_fn<void (APIServerConnection::*)()> (std::tr1::shared_ptr<APIServerConnection>)>>'
    handler_(arg1_, arg2_);
```

Seems like missing header? but i cant find anything... 

edit: earlier errors were overlooked, starts with 
`/src/eve-server/apiserver/APIServerListener.cpp:45:99: error: ‘boost::asio::ip::tcp::acceptor’ {aka ‘class boost::asio::basic_socket_acceptor<boost::asio::ip::tcp>’} has no member named ‘get_io_service’`

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -88,6 +88,9 @@ ELSEIF( CMAKE_COMPILER_IS_GNUCXX )
   SET( CMAKE_CXX_FLAGS "-std=c++0x ${CMAKE_CXX_FLAGS}" )
 ENDIF( MSVC )
 
+#Switching to C++11
+SET( CMAKE_CXX_FLAGS "-std=gnu++11 -march=native -Wno-deprecated-declarations -Werror=return-type -fabi-version=14 ${CMAKE_CXX_FLAGS}" )
+
 #
 # Try to detemine the exact source version using SCM.
 #

--- a/src/eve-server/apiserver/APIServer.cpp
+++ b/src/eve-server/apiserver/APIServer.cpp
@@ -104,7 +104,7 @@ std::string& APIServer::url()
 
 void APIServer::Run()
 {
-    _ioThread = std::unique_ptr<boost::asio::detail::thread>(new boost::asio::detail::thread(std::tr1::bind(&APIServer::RunInternal, this)));
+    _ioThread = std::unique_ptr<boost::asio::detail::thread>(new boost::asio::detail::thread(std::bind(&APIServer::RunInternal, this)));
 }
 
 void APIServer::Stop()
@@ -115,7 +115,7 @@ void APIServer::Stop()
 
 void APIServer::RunInternal()
 {
-    _io = std::unique_ptr<boost::asio::io_service>(new boost::asio::io_service());
+    _io = std::unique_ptr<boost::asio::io_context>(new boost::asio::io_context());
     _listener = std::unique_ptr<APIServerListener>(new APIServerListener(*_io));
     _io->run();
 }

--- a/src/eve-server/apiserver/APIServer.h
+++ b/src/eve-server/apiserver/APIServer.h
@@ -61,7 +61,7 @@ private:
     void RunInternal();
 
     std::unique_ptr<boost::asio::detail::thread> _ioThread;
-    std::unique_ptr<boost::asio::io_service> _io;
+    std::unique_ptr<boost::asio::io_context> _io;
     std::unique_ptr<APIServerListener> _listener;
     std::string _url;
     std::string _basePath;

--- a/src/eve-server/apiserver/APIServerConnection.cpp
+++ b/src/eve-server/apiserver/APIServerConnection.cpp
@@ -68,7 +68,7 @@ boost::asio::const_buffers_1 APIServerConnection::_responseNotFound = boost::asi
 "</html>"
 , 1187);
 
-APIServerConnection::APIServerConnection(boost::asio::io_service& io)
+APIServerConnection::APIServerConnection(boost::asio::io_context& io)
     : _socket(io)
 {
 }
@@ -81,7 +81,7 @@ boost::asio::ip::tcp::socket& APIServerConnection::socket()
 void APIServerConnection::Process()
 {
     // receive all HTTP headers from the client
-    boost::asio::async_read_until(_socket, _buffer, "\r\n\r\n", std::tr1::bind(&APIServerConnection::ProcessHeaders, shared_from_this()));
+    boost::asio::async_read_until(_socket, _buffer, "\r\n\r\n", std::bind(&APIServerConnection::ProcessHeaders, shared_from_this()));
 }
 
 void APIServerConnection::ProcessHeaders()
@@ -183,7 +183,7 @@ void APIServerConnection::ProcessHeaders()
             sLog.Debug("        ", "%d: param = %s,  value = %s", i, cur->first.c_str(), cur->second.c_str() );
 
         // first we have to send the responseOK, then our actual result
-        boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::SendXML, shared_from_this()));
+        boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::bind(&APIServerConnection::SendXML, shared_from_this()));
         //// DUPLICATE
     }
     else if (post_chk_str.compare("POST") == 0)
@@ -285,7 +285,7 @@ void APIServerConnection::ProcessHeaders()
                 // Call boost::asio::async_read() and feed it the # of bytes from step 1) to get the POST data
                 // The 'CompleteCondition' for THIS boost::asio::async_read, a parameter that specifies when to stop reading,
                 // is transfer_exactly(contentLength), where contentLength is the # of bytes we just recovered from the "Content-Length" header
-                boost::asio::async_read(_socket, _postBuffer, boost::asio::transfer_exactly(postDataBytes), std::tr1::bind(&APIServerConnection::ProcessPostData, shared_from_this()));
+                boost::asio::async_read(_socket, _postBuffer, boost::asio::transfer_exactly(postDataBytes), std::bind(&APIServerConnection::ProcessPostData, shared_from_this()));
             }
 
             _xmlData = sAPIServer.GetXML(&m_apiCommandCall);
@@ -305,7 +305,7 @@ void APIServerConnection::ProcessHeaders()
                 sLog.Debug("        ", "%d: param = %s,  value = %s", i, cur->first.c_str(), cur->second.c_str() );
 
             // first we have to send the responseOK, then our actual result
-            boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::SendXML, shared_from_this()));
+            boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::bind(&APIServerConnection::SendXML, shared_from_this()));
             //// DUPLICATE
         }
         else
@@ -313,7 +313,7 @@ void APIServerConnection::ProcessHeaders()
             // Call boost::asio::async_read() and feed it the # of bytes from step 1) to get the POST data
             // The 'CompleteCondition' for THIS boost::asio::async_read, a parameter that specifies when to stop reading,
             // is transfer_exactly(contentLength), where contentLength is the # of bytes we just recovered from the "Content-Length" header
-            boost::asio::async_read(_socket, _postBuffer, boost::asio::transfer_exactly(postDataBytes), std::tr1::bind(&APIServerConnection::ProcessPostData, shared_from_this()));
+            boost::asio::async_read(_socket, _postBuffer, boost::asio::transfer_exactly(postDataBytes), std::bind(&APIServerConnection::ProcessPostData, shared_from_this()));
         }
     }
     else
@@ -338,7 +338,7 @@ void APIServerConnection::ProcessPostData()
     if( request.compare( "" ) == 0 )
     {
         sLog.Error("APIServerConnection::ProcessPostData()", "POST data block is COMPLETELY EMPTY!!" );
-        boost::asio::async_write(_socket, _responseNoContent, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::Close, shared_from_this()));
+        boost::asio::async_write(_socket, _responseNoContent, boost::asio::transfer_all(), std::bind(&APIServerConnection::Close, shared_from_this()));
         //NotFound();
         return;
     }
@@ -383,22 +383,22 @@ void APIServerConnection::ProcessPostData()
         sLog.Debug("        ", "%d: param = %s,  value = %s", i, cur->first.c_str(), cur->second.c_str() );
 
     // first we have to send the responseOK, then our actual result
-    boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::SendXML, shared_from_this()));
+    boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::bind(&APIServerConnection::SendXML, shared_from_this()));
 }
 
 void APIServerConnection::SendXML()
 {
-    boost::asio::async_write(_socket, boost::asio::buffer(*_xmlData, _xmlData->size()), boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, boost::asio::buffer(*_xmlData, _xmlData->size()), boost::asio::transfer_all(), std::bind(&APIServerConnection::Close, shared_from_this()));
 }
 
 void APIServerConnection::NotFound()
 {
-    boost::asio::async_write(_socket, _responseNotFound, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, _responseNotFound, boost::asio::transfer_all(), std::bind(&APIServerConnection::Close, shared_from_this()));
 }
 
 void APIServerConnection::Redirect()
 {
-    boost::asio::async_write(_socket, _responseRedirectBegin, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::RedirectLocation, shared_from_this()));
+    boost::asio::async_write(_socket, _responseRedirectBegin, boost::asio::transfer_all(), std::bind(&APIServerConnection::RedirectLocation, shared_from_this()));
 }
 
 void APIServerConnection::RedirectLocation()
@@ -407,12 +407,12 @@ void APIServerConnection::RedirectLocation()
     std::stringstream url;
     url << APIServer::FallbackURL << _service;// << "/" << _id;
     _redirectUrl = url.str();
-    boost::asio::async_write(_socket, boost::asio::buffer(_redirectUrl), boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::RedirectFinalize, shared_from_this()));
+    boost::asio::async_write(_socket, boost::asio::buffer(_redirectUrl), boost::asio::transfer_all(), std::bind(&APIServerConnection::RedirectFinalize, shared_from_this()));
 }
 
 void APIServerConnection::RedirectFinalize()
 {
-    boost::asio::async_write(_socket, _responseRedirectEnd, boost::asio::transfer_all(), std::tr1::bind(&APIServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, _responseRedirectEnd, boost::asio::transfer_all(), std::bind(&APIServerConnection::Close, shared_from_this()));
 }
 
 void APIServerConnection::Close()
@@ -425,7 +425,7 @@ bool APIServerConnection::starts_with(std::string& haystack, const char *const n
     return haystack.substr(0, strlen(needle)).compare(needle) == 0;
 }
 
-std::tr1::shared_ptr<APIServerConnection> APIServerConnection::create(boost::asio::io_service& io)
+std::tr1::shared_ptr<APIServerConnection> APIServerConnection::create(boost::asio::io_context& io)
 {
     return std::tr1::shared_ptr<APIServerConnection>(new APIServerConnection(io));
 }

--- a/src/eve-server/apiserver/APIServerConnection.h
+++ b/src/eve-server/apiserver/APIServerConnection.h
@@ -41,12 +41,12 @@
 class APIServerConnection : public std::tr1::enable_shared_from_this<APIServerConnection>
 {
 public:
-    static std::tr1::shared_ptr<APIServerConnection> create(boost::asio::io_service& io);
+    static std::tr1::shared_ptr<APIServerConnection> create(boost::asio::io_context& io);
     void Process();
     boost::asio::ip::tcp::socket& socket();
 
 private:
-    APIServerConnection(boost::asio::io_service& io);
+    APIServerConnection(boost::asio::io_context& io);
     void ProcessHeaders();
     void ProcessPostData();
     void SendXML();

--- a/src/eve-server/apiserver/APIServerListener.cpp
+++ b/src/eve-server/apiserver/APIServerListener.cpp
@@ -28,7 +28,7 @@
 #include "EVEServerConfig.h"
 #include "apiserver/APIServerListener.h"
 
-APIServerListener::APIServerListener(boost::asio::io_service& io)
+APIServerListener::APIServerListener(boost::asio::io_context& io)
 {
     _acceptor = new boost::asio::ip::tcp::acceptor(io, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), sConfig.net.apiServerPort));
     sLog.Log("API Server Init", "listening on port %u", (sConfig.net.apiServerPort));
@@ -42,8 +42,12 @@ APIServerListener::~APIServerListener()
 
 void APIServerListener::StartAccept()
 {
-    std::tr1::shared_ptr<APIServerConnection> connection = APIServerConnection::create(_acceptor->get_io_service());
-    _acceptor->async_accept(connection->socket(), std::tr1::bind(&APIServerListener::HandleAccept, this, connection));
+    boost::asio::executor e = _acceptor->get_executor();
+    boost::asio::execution_context &e_context = e.context();
+    boost::asio::io_context &context_instance = static_cast<boost::asio::io_context&>(e_context);
+
+    std::tr1::shared_ptr<APIServerConnection> connection = APIServerConnection::create(context_instance);
+    _acceptor->async_accept(connection->socket(), std::bind(&APIServerListener::HandleAccept, this, connection));
 }
 
 void APIServerListener::HandleAccept(std::tr1::shared_ptr<APIServerConnection> connection)

--- a/src/eve-server/apiserver/APIServerListener.h
+++ b/src/eve-server/apiserver/APIServerListener.h
@@ -41,7 +41,7 @@
 class APIServerListener
 {
 public:
-    APIServerListener(boost::asio::io_service& io);
+    APIServerListener(boost::asio::io_context& io);
     ~APIServerListener();
 
 private:

--- a/src/eve-server/imageserver/ImageServer.cpp
+++ b/src/eve-server/imageserver/ImageServer.cpp
@@ -180,7 +180,7 @@ std::string& ImageServer::url()
 
 void ImageServer::Run()
 {
-    _ioThread = std::shared_ptr<boost::asio::detail::thread>(new boost::asio::detail::thread(std::tr1::bind(&ImageServer::RunInternal, this)));
+    _ioThread = std::shared_ptr<boost::asio::detail::thread>(new boost::asio::detail::thread(std::bind(&ImageServer::RunInternal, this)));
 }
 
 void ImageServer::Stop()
@@ -191,7 +191,7 @@ void ImageServer::Stop()
 
 void ImageServer::RunInternal()
 {
-    _io = std::shared_ptr<boost::asio::io_service>(new boost::asio::io_service());
+    _io = std::shared_ptr<boost::asio::io_context>(new boost::asio::io_context());
     _listener = std::shared_ptr<ImageServerListener>(new ImageServerListener(*_io));
     _io->run();
 }

--- a/src/eve-server/imageserver/ImageServer.h
+++ b/src/eve-server/imageserver/ImageServer.h
@@ -70,7 +70,7 @@ private:
 
     std::tr1::unordered_map<uint32 /*accountID*/, std::tr1::shared_ptr<std::vector<char> > /*imageData*/> _limboImages;
     std::shared_ptr<boost::asio::detail::thread> _ioThread;
-    std::shared_ptr<boost::asio::io_service> _io;
+    std::shared_ptr<boost::asio::io_context> _io;
     std::shared_ptr<ImageServerListener> _listener;
     std::string _url;
     std::string _basePath;

--- a/src/eve-server/imageserver/ImageServerConnection.cpp
+++ b/src/eve-server/imageserver/ImageServerConnection.cpp
@@ -33,7 +33,7 @@ boost::asio::const_buffers_1 ImageServerConnection::_responseNotFound = boost::a
 boost::asio::const_buffers_1 ImageServerConnection::_responseRedirectBegin = boost::asio::buffer("HTTP/1.0 301 Moved Permanently\r\nLocation: ", 42);
 boost::asio::const_buffers_1 ImageServerConnection::_responseRedirectEnd = boost::asio::buffer("\r\n\r\n", 4);
 
-ImageServerConnection::ImageServerConnection(boost::asio::io_service& io)
+ImageServerConnection::ImageServerConnection(boost::asio::io_context& io)
     : _socket(io)
 {
 }
@@ -46,7 +46,7 @@ boost::asio::ip::tcp::socket& ImageServerConnection::socket()
 void ImageServerConnection::Process()
 {
     // receive all HTTP headers from the client
-    boost::asio::async_read_until(_socket, _buffer, "\r\n\r\n", std::tr1::bind(&ImageServerConnection::ProcessHeaders, shared_from_this()));
+    boost::asio::async_read_until(_socket, _buffer, "\r\n\r\n", std::bind(&ImageServerConnection::ProcessHeaders, shared_from_this()));
 }
 
 void ImageServerConnection::ProcessHeaders()
@@ -109,22 +109,22 @@ void ImageServerConnection::ProcessHeaders()
     }
 
     // first we have to send the responseOK, then our actual result
-    boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::SendImage, shared_from_this()));
+    boost::asio::async_write(_socket, _responseOK, boost::asio::transfer_all(), std::bind(&ImageServerConnection::SendImage, shared_from_this()));
 }
 
 void ImageServerConnection::SendImage()
 {
-    boost::asio::async_write(_socket, boost::asio::buffer(*_imageData, _imageData->size()), boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, boost::asio::buffer(*_imageData, _imageData->size()), boost::asio::transfer_all(), std::bind(&ImageServerConnection::Close, shared_from_this()));
 }
 
 void ImageServerConnection::NotFound()
 {
-    boost::asio::async_write(_socket, _responseNotFound, boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, _responseNotFound, boost::asio::transfer_all(), std::bind(&ImageServerConnection::Close, shared_from_this()));
 }
 
 void ImageServerConnection::Redirect()
 {
-    boost::asio::async_write(_socket, _responseRedirectBegin, boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::RedirectLocation, shared_from_this()));
+    boost::asio::async_write(_socket, _responseRedirectBegin, boost::asio::transfer_all(), std::bind(&ImageServerConnection::RedirectLocation, shared_from_this()));
 }
 
 void ImageServerConnection::RedirectLocation()
@@ -133,12 +133,12 @@ void ImageServerConnection::RedirectLocation()
     std::stringstream url;
     url << ImageServer::FallbackURL << _category << "/" << _id << "_" << _size << "." << extension;
     _redirectUrl = url.str();
-    boost::asio::async_write(_socket, boost::asio::buffer(_redirectUrl), boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::RedirectFinalize, shared_from_this()));
+    boost::asio::async_write(_socket, boost::asio::buffer(_redirectUrl), boost::asio::transfer_all(), std::bind(&ImageServerConnection::RedirectFinalize, shared_from_this()));
 }
 
 void ImageServerConnection::RedirectFinalize()
 {
-    boost::asio::async_write(_socket, _responseRedirectEnd, boost::asio::transfer_all(), std::tr1::bind(&ImageServerConnection::Close, shared_from_this()));
+    boost::asio::async_write(_socket, _responseRedirectEnd, boost::asio::transfer_all(), std::bind(&ImageServerConnection::Close, shared_from_this()));
 }
 
 void ImageServerConnection::Close()
@@ -151,7 +151,7 @@ bool ImageServerConnection::starts_with(std::string& haystack, const char *const
     return haystack.substr(0, strlen(needle)).compare(needle) == 0;
 }
 
-std::tr1::shared_ptr<ImageServerConnection> ImageServerConnection::create(boost::asio::io_service& io)
+std::tr1::shared_ptr<ImageServerConnection> ImageServerConnection::create(boost::asio::io_context& io)
 {
     return std::tr1::shared_ptr<ImageServerConnection>(new ImageServerConnection(io));
 }

--- a/src/eve-server/imageserver/ImageServerConnection.h
+++ b/src/eve-server/imageserver/ImageServerConnection.h
@@ -39,12 +39,12 @@
 class ImageServerConnection : public std::tr1::enable_shared_from_this<ImageServerConnection>
 {
 public:
-    static std::tr1::shared_ptr<ImageServerConnection> create(boost::asio::io_service& io);
+    static std::tr1::shared_ptr<ImageServerConnection> create(boost::asio::io_context& io);
     void Process();
     boost::asio::ip::tcp::socket& socket();
 
 private:
-    ImageServerConnection(boost::asio::io_service& io);
+    ImageServerConnection(boost::asio::io_context& io);
     void ProcessHeaders();
     void SendImage();
     void NotFound();

--- a/src/eve-server/imageserver/ImageServerListener.cpp
+++ b/src/eve-server/imageserver/ImageServerListener.cpp
@@ -45,11 +45,11 @@ void ImageServerListener::StartAccept()
     boost::asio::execution_context &e_context = e.context();
     boost::asio::io_context &context_instance = static_cast<boost::asio::io_context&>(e_context);
 
-    std::shared_ptr<ImageServerConnection> connection = ImageServerConnection::create(context_instance);
+    std::tr1::shared_ptr<ImageServerConnection> connection = ImageServerConnection::create(context_instance);
     _acceptor->async_accept(connection->socket(), std::bind(&ImageServerListener::HandleAccept, this, connection));
 }
 
-void ImageServerListener::HandleAccept(std::shared_ptr<ImageServerConnection> connection)
+void ImageServerListener::HandleAccept(std::tr1::shared_ptr<ImageServerConnection> connection)
 {
     connection->Process();
     StartAccept();

--- a/src/eve-server/imageserver/ImageServerListener.cpp
+++ b/src/eve-server/imageserver/ImageServerListener.cpp
@@ -28,10 +28,9 @@
 #include "EVEServerConfig.h"
 #include "imageserver/ImageServerListener.h"
 
-ImageServerListener::ImageServerListener(boost::asio::io_service& io)
+ImageServerListener::ImageServerListener(boost::asio::io_context& io)
 {
     _acceptor = new boost::asio::ip::tcp::acceptor(io, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), sConfig.net.imageServerPort));
-    sLog.Log("Image Server Init", "listening on port %i", (sConfig.net.imageServerPort));
     StartAccept();
 }
 
@@ -42,11 +41,15 @@ ImageServerListener::~ImageServerListener()
 
 void ImageServerListener::StartAccept()
 {
-    std::tr1::shared_ptr<ImageServerConnection> connection = ImageServerConnection::create(_acceptor->get_io_service());
-    _acceptor->async_accept(connection->socket(), std::tr1::bind(&ImageServerListener::HandleAccept, this, connection));
+    boost::asio::executor e = _acceptor->get_executor();
+    boost::asio::execution_context &e_context = e.context();
+    boost::asio::io_context &context_instance = static_cast<boost::asio::io_context&>(e_context);
+
+    std::shared_ptr<ImageServerConnection> connection = ImageServerConnection::create(context_instance);
+    _acceptor->async_accept(connection->socket(), std::bind(&ImageServerListener::HandleAccept, this, connection));
 }
 
-void ImageServerListener::HandleAccept(std::tr1::shared_ptr<ImageServerConnection> connection)
+void ImageServerListener::HandleAccept(std::shared_ptr<ImageServerConnection> connection)
 {
     connection->Process();
     StartAccept();

--- a/src/eve-server/imageserver/ImageServerListener.h
+++ b/src/eve-server/imageserver/ImageServerListener.h
@@ -41,7 +41,7 @@
 class ImageServerListener
 {
 public:
-    ImageServerListener(boost::asio::io_service& io);
+    ImageServerListener(boost::asio::io_context& io);
     ~ImageServerListener();
 
 private:

